### DISCUSSION TITLE
Write new state to Home Assistant after sending command(s) to Pura & prior to requesting a refresh / Don't poll Pura API immediately after sending command(s)

### DISCUSSION THREAD

**Discussion by RobertD502** [natekspencer/hacs-pura#24] [Ideas]
I noticed that you don't write the ["new" state to home assistant](https://github.com/natekspencer/hacs-pura/blob/8c7c3ef368f2f3423f2d41182649c820112906e6/custom_components/pura/number.py#L85-L105) after sending a command to pura's API. As a result, for example, if my current intensity is 5 and I bump it up to 8 with the slider number entity, during a page refresh it reverts back to 5 until the call to the pura API returns updated coordinator data.

One way around this is to manually write the new state to the coordinator object prior to `async_request_refresh()`. For example, with the scent intensity number entity you'd add the following prior to the `async_request_refresh()` function:

```python
self._intensity_data["intensity"] = value
self.async_write_ha_state()
# Wait until refreshing data as the Pura API may return the old intensity if polled immediately after sending a command
await asyncio.sleep(1)
await self.coordinator.async_request_refresh()
```

**If not writing the state and switching from an intensity of 5 to 8, notice the number entity reverts back to 5 when the page is refreshed:**

https://github.com/natekspencer/pypura/assets/52541649/048810a7-d869-4234-b9d4-4fe8d32e424b

**When writing the state - notice the number entity remains at the value it was just set to when refreshing the page:**

https://github.com/natekspencer/pypura/assets/52541649/88b76c87-d577-4e2e-918c-f86b981869f7

Although the reverting of values is most commonly seen if the HA UI is refreshed (the UI uses the old value from the coordinator data during a page refresh), this can also happen if the Pura API is polled prior to the new value being reflected in the returned data.

1) `await asyncio.sleep(1)` circumvents the polling issue by not polling the API until a second after a command is sent in order to ensure the returned Pura API data is up-to-date.
2) Writing the command value to the coordinator prior to requesting a refresh of the data circumvents the issue seen if a user refreshes the HA UI after sending a command to Pura.

It would also be worthwhile doing this for the `select`, `switch`, and `light` entities since they also involve sending commands.

**Comment by natekspencer** [comment]
Thanks for this suggestion! In general, I think it's a good idea with a couple of caveats. The coordinator may still poll between the sleep and the requested refresh, so it's very possible that it will still exhibit this issue. Additionally, the forced 1-second sleep may not always guarantee that the Pura API is updated for the return

**Reply by RobertD502** [reply]
Touché. I'll have to look into it, but it may be possible to natively add a step where the coordinator is paused prior to sleeping and then the request refresh function would bring it back into the scheduled coordinator updates.

Regarding `the forced 1-second sleep may not always guarantee that the Pura API is updated for the return`: That is very true. We are limited by how fast their API updates and in some instances it may still return old data - it essentially becomes a battle of how long of a sleep is too long in order to be confident such a scenario is covered, but that is up to you to decide and even if it is something you'd want in your code.

**Reply by natekspencer** [reply]
I had to do something similar in the Litter-Robot integration before I was able to migrate it to websockets. I essentially created an assumed state that expired after the refresh callback: https://github.com/home-assistant/core/pull/59805/files

I'll admit, I haven't been motivated to implement a similar thing here because I primarily call the timer service and don't interact via the UI 😅

**Reply by RobertD502** [reply]
I'll be honest, I never use the UI myself on any of my integration since I prefer for things to be automated. However, I know that not everyone is the same so I try to make sure things are smooth for heavy UI users and prevent it from becoming an issue ticket that I have to deal with at a later time 😆.

**Reply by natekspencer** [reply]
I get that. This is also one of those where I've never actually had a device with only 1 fragrance installed, so never even considered it or knew what the data looked like coming back from the API haha. (Sorry this is unrelated, I was just looking at your PR in relation to the timer and got my wires crossed, lol)

**Reply by natekspencer** [reply]
A lot of it for me is also time permitting. I'll admit I put this integration together pretty quickly and have a lot of other integrations I work on myself and is why I still consider this a pre-1.0.0 release haha.

**Comment by RobertD502** [comment]
> I get that. This is also one of those where I've never actually had a device with only 1 fragrance installed, so never even considered it or knew what the data looked like coming back from the API haha. (Sorry this is unrelated, I was just looking at your PR in relation to the timer and got my wires crossed, lol)

Ah, we just got 3 devices ~3 weeks ago and wanted to see how well the scents would work so only installed one in each and kept the others for future use, which is the only reason why I was able to track down what was causing the issue the user reported.

**Reply by RobertD502** [reply]
Whoops it started a new comment thread.

**Reply by natekspencer** [reply]
The other fun this is that you can't actually remove a fragrance without replacing it with a new one either. So once you have two, there is no going back. Unless you completely remove the device first and start over..

**Reply by RobertD502** [reply]
So if you selected slot 2 and that slot is empty/you physically took the vial out, it would still attempt to heat slot 2 even if no vial is present? Seems.....not well thought out. Is it possible to rescan the same vial and place it in slot 2 in order to reset the `wearingTime`? If not, that would mean the automatic selection via the `start_timer` service will only select slot 1 if the time is less than slot 2 - In the event that your slot 1 has plenty of fragrance and slot 2 ran out of fragrance (you ran it at a high intensity), but slot 2 has a shorter `wearingTime`, it would pointlessly attempt to use slot 2 even though it is technically "empty".

**Reply by natekspencer** [reply]
The Pura 4 device may be able to do it automatically. It's the Pura 3 device that has this specific issue. And I've never actually just pulled it out without putting another one in. I just let it stay in there until I replace both of them.

**Reply by RobertD502** [reply]
What's your experience with the scents? I find that the current scents that I have don't really do much on an intensity of 5 - I have 12 foot ceilings so I'm starting to think it may be due to them just not being powerful enough at a modest intensity setting. However, running at a 10 would deplete the fragrance super fast.

**Reply by RobertD502** [reply]
Side note: not sure if you care about knowing download stats, but if you add a workflow to zip up your releases (and also change the hacs settings to zip release), you'll be able to see download stats per release (either via your repo page on HACS or manually using the github API to check download counts on releases) - this includes users that have turned off analytics in Home Assistant.

[Example](https://github.com/RobertD502/home-assistant-flair/blob/main/.github/workflows/release.yml)
